src/MeshlessFV/MeshlessFVSimulation.cpp

```diff
@@ -486,31 +548,54 @@ void MeshlessFVSimulation<ndim>::PostInitialConditionsSetup(void)
   rebuild_tree = true;
   mfvneib->BuildTree(rebuild_tree, 0, ntreebuildstep, ntreestockstep, mfv->Ntot,
                      mfv->Nhydromax, timestep, partdata, mfv);
+#ifdef MPI_PARALLEL
+  mfvneib->InitialiseCellWorkCounters();
+#endif
 
   // For Eigenvalue MAC, need non-zero values
   for (i=0; i<mfv->Nhydro; i++) partdata[i].gpot = big_number;
 
   // Calculate all hydro properties
   mfvneib->UpdateAllProperties(mfv->Nhydro ,mfv->Ntot, partdata, mfv, nbody, simbox);
 
+#ifdef MPI_PARALLEL
+  mpicontrol->UpdateAllBoundingBoxes(mfv->Nhydro, mfv, mfv->kernp);
+#endif
+
   // Search ghost particles
   mfvneib->SearchBoundaryGhostParticles((FLOAT) 0.0, simbox, mfv);
   // Update pointer in case there has been a reallocation
   partdata = mfv->GetMeshlessFVParticleArray();
   mfvneib->BuildGhostTree(true, 0, ntreebuildstep, ntreestockstep, mfv->Ntot,
                           mfv->Nhydromax, timestep, partdata, mfv);
+#ifdef MPI_PARALLEL
+  mpicontrol->UpdateAllBoundingBoxes(mfv->Nhydro + mfv->NPeriodicGhost, mfv, mfv->kernp);
+  MpiGhosts->SearchGhostParticles((FLOAT) 0.0, simbox, mfv);
+  // Update pointer in case there has been a reallocation
+  partdata = mfv->GetMeshlessFVParticleArray();
+  mfvneib->BuildMpiGhostTree(true, 0, ntreebuildstep, ntreestockstep, mfv->Ntot,
+                             mfv->Nhydromax, timestep, partdata, mfv);
+#endif
 
   // Update neighbour tree
   rebuild_tree = true;
```

Comment:
i'm focusing on the main loop for now - performance is not so important in PostInitialConditionsSetup (but I will go back to it later)